src/conduits/RwaInputConduit3.sol

```diff
@@ -0,0 +1,296 @@
+
+    /**
+     * @notice Grants `usr` push access to this contract.
+     * @param usr The user address.
+     */
+    function mate(address usr) external auth {
+        may[usr] = 1;
+        emit Mate(usr);
+    }
+
+    /**
+     * @notice Revokes `usr` push access from this contract.
+     * @param usr The user address.
+     */
+    function hate(address usr) external auth {
+        may[usr] = 0;
+        emit Hate(usr);
+    }
+
+    /*//////////////////////////////////
+               Administration
+    //////////////////////////////////*/
+
+    /**
+     * @notice Updates a contract parameter.
+     * @param what The changed parameter name. `"quitTo", "to"`
+     * @param data The new value of the parameter.
+     */
+    function file(bytes32 what, address data) external auth {
+        if (what == "quitTo") {
+            require(data != address(0), "RwaInputConduit3/invalid-quit-to-address");
+            quitTo = data;
+        } else if (what == "to") {
+            require(data != address(0), "RwaInputConduit3/invalid-to-address");
+            to = data;
+        } else {
+            revert("RwaInputConduit3/unrecognised-param");
+        }
+
+        emit File(what, data);
+    }
+
+    /*//////////////////////////////////
+               Operations
+    //////////////////////////////////*/
+
+    /**
+     * @notice Swaps the GEM balance of this contract into DAI through the PSM and push it into the `to` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     */
+    function push() external isMate {
+        _doPush(gem.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Swaps the specified amount of GEM into DAI through the PSM and push it into the `to` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     * @param amt Gem amount.
+     */
+    function push(uint256 amt) external isMate {
+        _doPush(amt);
+    }
+
+    /**
+     * @notice Flushes out any GEM balance to `quitTo` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     */
+    function quit() external isMate {
+        _doQuit(gem.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Flushes out specific amount of GEM balance to `quitTo` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     * @param amt Gem amount.
+     */
+    function quit(uint256 amt) external isMate {
+        _doQuit(amt);
+    }
+
+    /**
+     * @notice Flushes out all outstanding DAI balance to `usr` address.
+     * @dev Can only be called by the admin
+     * @param usr Destination address.
+     */
+    function quitDai(address usr) external auth {
+        dai.transfer(usr, dai.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Calculate amount of DAI received for swapping GEM through PSM.
+     * @param gemAmt GEM amount.
+     */
+    function gemToDai(uint256 gemAmt) external view returns (uint256) {
```

Comment:
is this function needed? I don't think it means much as the psm fees are treated as part of the payback, so even if the resulting dai is smaller than the USDC amount the operator shouldn't care.

Reply:
Discussing this internally we will probably change this function.

It should exist as a convenience method for those running the vaults to answer one question:

> How much USDC do I need to be able to get X amount of DAI?

Because of the differences in precision and also the PSM fees, this is not trivially done by non-expert people.

So we will add a function named `requiredGemAmt(uint256 daiWad) view returns (uint256 gemAmt)` that handles that conversion for the user, taking the current PSM fees into consideration.

Reply:
Done

Reply:
Ok, but do we really expect the party taking the loan to repay more usdc just because `tin` is bigger? I might be mistaking but I assume that's not the deal.
I mean I assume that even if there is a large `tin` and we end up with less dai closing the debt, the psm fee will be regarded as part of the returned funds to the protocol, and the urn debt would eventually be needed to close with `grab`.
Is that the intention? If so then we might want to remove `requiredGemAmt`.

Reply:
Resolving this as it was discussed here - https://github.com/makerdao/mip21-toolkit/pull/21#discussion_r974059953
I still think most of the helper functions will not be used externally (except maybe `requiredDaiAmt` in the output conduit).